contracts/validator-manager/PoSValidatorManager.sol

```diff
@@ -357,112 +357,62 @@ abstract contract PoSValidatorManager is
         PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
 
         Delegator memory delegator = $._delegatorStakes[delegationID];
+        bytes32 validationID = delegator.validationID;
         Validator memory validator = getValidator(delegator.validationID);
 
         // Ensure the delegator is pending added. Since anybody can call this function once
         // delegator registration has been initialized, we need to make sure that this function is only
         // callable after that has been done.
-        if ($._delegatorStakes[delegationID].status != DelegatorStatus.PendingAdded) {
+        if (delegator.status != DelegatorStatus.PendingAdded) {
             revert InvalidDelegatorStatus();
         }
 
+        // In the case where the validator has completed its validation period, we can no
+        // longer stake and should move our status directly to completed and return the stake.
+        if (validator.status == ValidatorStatus.Completed) {
+            return _completeEndDelegation(delegationID);
+        }
+
         // In the case where the validator has initiated ending its validation period, we can no
         // longer stake, but the validator's weight may not have been removed from the p-chain yet,
         // so we have to wait for confirmation that the validator has been removed before returning the stake.
+        uint64 delegatorStartTime;
         if (validator.status == ValidatorStatus.PendingRemoved) {
-            delegator.status = DelegatorStatus.PendingRemoved;
-            delegator.startedAt = validator.endedAt;
-            delegator.endingNonce = validator.messageNonce;
-
-            // Write back the delegator
-            $._delegatorStakes[delegationID] = delegator;
-
-            emit DelegatorRemovalInitialized({
-                delegationID: delegationID,
-                validationID: delegator.validationID,
-                endTime: validator.endedAt
-            });
-
-            return;
-        }
+            delegatorStartTime = validator.endedAt;
```

Comment:
Could you explain this case again?

It seems incorrect to me to not require a valid message here. If a delegator starts to register, and then the validator starts to exit, we should still require proof that the delegator's message has been delivered to the P-Chain in order to mark them as `active` right?

I guess marking them as `active` doesn't have an effect either way since their end time is locked in and is no later than their start time, but it may be cleaner and more intuitive to have these two cases handled exactly the same, and then handle the possibility of `start time > end time` when removing the delegator

Reply:
Hmm, yeah I think I like the handling of `start time > end time` separately.